### PR TITLE
Feature/aql subquery operations stack enumerate list

### PR DESCRIPTION
**Pull request by hkernbach**
EnumerateList Executor

Mergeable after Jenkins is green + Todos are all done. 

Todos: 
- [x] Optimize internal skip Loop
- [x] Write more C++ tests (skipRowsRange)
- [x] Add a C++ failure test

### CODE REVIEW

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8275/

**Comment by hkernbach** [comment]
Next Jenkins Run:
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8369/

**Comment by hkernbach** [comment]
Made skipping better (prev. run was green)
Jenkins: http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8398/

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8490/

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8500/

**Comment by hkernbach** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8526/

**Comment by hkernbach** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8530/

File: tests/Aql/ExecutorTestHelper.h [hunk 1]
```diff
@@ -145,6 +151,13 @@ struct ExecutorTestHelper {
 
     SharedAqlItemBlockPtr expectedOutputBlock =
         buildBlock<outputColumns>(itemBlockManager, std::move(_output));
+
+    if (result == nullptr) {
+      EXPECT_EQ(expectedOutputBlock, nullptr);
+      EXPECT_EQ(_emptyOutput, true);
+      return;
+    }
```
Comment (maierlars) [inline tests/Aql/ExecutorTestHelper.h]: What happens if `result` is not `nullptr` but `_emptyOutput == true`. I think in that case `testOutputBlock` does a `nullptr`-access.

File: arangod/Aql/EnumerateListExecutor.cpp [hunk 2]
```diff
@@ -70,84 +72,149 @@ RegisterId EnumerateListExecutorInfos::getOutputRegister() const noexcept {
...
-      }
-      initialize();
-      std::tie(_rowState, _currentRow) = _fetcher.fetchRow();
-      if (_rowState == ExecutionState::WAITING) {
-        return {_rowState, NoStats{}};
-      }
-    }
+  TRI_ASSERT(false);
+  THROW_ARANGO_EXCEPTION(TRI_ERROR_NOT_IMPLEMENTED);
+}
```
Comment (maierlars) [inline arangod/Aql/EnumerateListExecutor.cpp]: Afaik we agreed on removing the old methods.

File: arangod/Aql/EnumerateListExecutor.cpp [hunk 3]
```diff
@@ -70,84 +72,149 @@ RegisterId EnumerateListExecutorInfos::getOutputRegister() const noexcept {
...
+
+std::tuple<ExecutorState, NoStats, AqlCall> EnumerateListExecutor::produceRows(
+    AqlItemBlockInputRange& inputRange, OutputAqlItemRow& output) {
+  AqlCall upstreamCall{};
+  upstreamCall.fullCount = output.getClientCall().fullCount;
+
+  if (!inputRange.hasDataRow()) {
+    return {inputRange.upstreamState(), NoStats{}, upstreamCall};
+  }
```
Comment (maierlars) [inline arangod/Aql/EnumerateListExecutor.cpp]: I am always confused with that... what if the output was full and thus the current array has elements left. In that case I think you would receive the same input row containing the data again as you only consume the row when you wrote the array completely?

File: arangod/Aql/EnumerateListExecutor.cpp [hunk 4]
```diff
@@ -70,84 +72,149 @@ RegisterId EnumerateListExecutorInfos::getOutputRegister() const noexcept {
...
-      if (_rowState == ExecutionState::HASMORE) {
-        continue;
-      } else {
-        return {_rowState, NoStats{}};
-      }
-    } else {
-      bool mustDestroy;
-      AqlValue innerValue = getAqlValue(inputList, _inputArrayPosition, mustDestroy);
-      AqlValueGuard guard(innerValue, mustDestroy);
+    TRI_ASSERT(_inputArrayPosition <= _inputArrayLength);
```
Comment (maierlars) [inline arangod/Aql/EnumerateListExecutor.cpp]:
```suggestion
    TRI_ASSERT(_inputArrayPosition < _inputArrayLength);
```

File: arangod/Aql/EnumerateListExecutor.cpp [hunk 5]
```diff
@@ -70,84 +72,149 @@ RegisterId EnumerateListExecutorInfos::getOutputRegister() const noexcept {
...
+  while (inputRange.hasDataRow() && call.shouldSkip()) {
+    if (_inputArrayLength == _inputArrayPosition) {
+      // we reached either the end of an array
+      // or are in our first loop iteration
+      initializeNewRow(inputRange);
+      continue;
+    }
+    // auto const& [state, input] = inputRange.peekDataRow();
+
+    TRI_ASSERT(_inputArrayPosition <= _inputArrayLength);
```
Comment (maierlars) [inline arangod/Aql/EnumerateListExecutor.cpp]:
```suggestion
    TRI_ASSERT(_inputArrayPosition < _inputArrayLength);
```

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8561/

**Comment by hkernbach** [comment]
and again http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8591/

**Pull request merged**